text/0002-signature-based-completions.md

```diff
@@ -0,0 +1,216 @@
+
+    > ls "some dir/
+
+doesn't parse successfully, since the string is unterminated. That being said, the parser knows that
+it's attempting to parse a string, so it should be able to convey that to the caller so that they
+can take better action on this error state.
+
+After lite parse, we need to follow up with a full parse to figure out which shape is under the
+cursor. If this succeeds, we have a shape. If it errors, we provide no completion suggestions.
+
+The second technical aspect of this completion engine is to make it store a registry of completers.
+This registry would map completion locations and syntax shapes to a list of completers. For example,
+we may have a registry like this:
+
+  | completion location | syntax shape | completers |
+  |---------------------|--------------|------------|
+  | command             | string       | filename, internal command, external command |
+  | flag/arg value      | string       | filename |
+  | flag/arg value      | path         | filename |
+  | flag name           | string       | flag name |
+  | variable            | string       | variable name |
+
+A completer would be described by the following trait:
+
+```rust
+trait Completer {
+    fn complete(line: &str, span: Span, context: completion::Context) -> Vec<Suggestion>;
+}
+```
+
+Concrete implementations are expected to take the given span for the line and provide some
+completions based on the partial data. The completion context will allow completers access to the
+nushell context, so they can discover variable names, environment variables, and so on. For example,
+a simple completer may look like this:
+
+```rust
+impl Completer for MyCompleter {
+    fn complete(line: &str, span: Span, context: completion::Context) -> Vec<Suggestion> {
+        if span.string(line) == "hello" {
+            vec![Suggestion::from("hello world")]
+        } else {
+            Vec::new()
+        }
+    }
+}
+```
+
+Completers don't have to worry about replacement, they simply say what the span should be replaced
+with. It's uncertain whether or not being able to replace a different span is necessary, so the
+completion engine will start with the assumption that it is unnecessary.
+
+The completion engine should deal with special needs of various shapes. For example, a string with
+spaces will need to be quoted. Completers should not have to worry about this, and leave it up to
+the engine. For example, when completing `hello` to `hello world`, like above, the completion engine
+would need to add the appropriate quoting when replacing the span. Continuing with this example,
+
+    git co hello
+
+could be completed to
+
+    git co "hello world"
+
+The span will never include quoting characters and other syntactical elements that aren't considered
+part of the value. Since completers have access to the full line, they can still consider those
+elements, but ideally completers never have to concern themselves with syntax.
+
```

Comment:
Does this completion include escaping characters? For instance, on Windows one can type `cd "Program Files"` and it'll work or also in powershell ```cd Program` Files``` and that will work too, not to mention `cd PROGRA~1` which is a windows `short file name`. As I recall escaping isn't a thing in nu.

Also, kind of related, will file name or folders with spaces in the be quoted with double quotes or single quotes? I've seen both ways. Maybe they're interchangeable.

Reply:
> Does this completion include escaping characters?

I didn't go into detail here, but good call. I think the idea of escaping/quoting makes sense in the context of this RFCs, so I'll go ahead and write up something on that.

> Also, kind of related, will file name or folders with spaces in the be quoted with double quotes or single quotes? I've seen both ways. Maybe they're interchangeable.

They should be interchangeable. I've recently updated our current completion code so that it chooses whichever quote works. For example, if you have a filename with a double quote in it, the quoting code would have to use a single quote. If you have both a single and double quote, it would use a backtick (format string). If you have all three, there's not really anything we can do right now 😬